devito/ir/clusters/algorithms.py

```diff
@@ -206,7 +206,13 @@ def guard(clusters):
                     k = d
                 else:
                     dims = pull_dims(cd.condition)
-                    k = max(dims, default=d, key=lambda i: c.ispace.index(i))
+                    # If `cd` uses more dimensions than the ispace,
+                    # stay under parent
+                    if (not dims.issubset(set(c.ispace.dimensions)) and
+                       cd.parent in dims):
+                        k = cd.parent
+                    else:
+                        k = max(dims, default=d, key=lambda i: c.ispace.index(i))
```

Comment:
Solution 2 with False flag

```
                # If `cd` uses, as condition, an arbitrary SymPy expression, then
                # we must ensure to nest it inside the last of the Dimensions
                # appearing in `cd.condition`
                if cd._factor is not None:
                    k = d
                else:
                    dims = pull_dims(cd.condition, flag=False)
                    k = max(dims, default=d, key=lambda i: c.ispace.index(i))
```

Reply:
Do you also consider the second solution not good enough?

Reply:
It's not clear _why_ passing `flag=False` would fix the issue here?